sensei/Catalyst2AnalysisAdaptor.cxx

```diff
@@ -0,0 +1,150 @@
+#include <svtkDataSet.h>
+#include <svtkMultiBlockDataSet.h>
+#include <svtkPointData.h>
+#include <svtkRange.h>
+
+namespace sensei
+{
+
+//-----------------------------------------------------------------------------
+senseiNewMacro(Catalyst2AnalysisAdaptor);
+
+//-----------------------------------------------------------------------------
+Catalyst2AnalysisAdaptor::Catalyst2AnalysisAdaptor()
+{
+  this->Initialize();
+}
+
+//-----------------------------------------------------------------------------
+Catalyst2AnalysisAdaptor::~Catalyst2AnalysisAdaptor()
+{
+  this->Finalize();
+}
+
+//-----------------------------------------------------------------------------
+void Catalyst2AnalysisAdaptor::Initialize() {}
+
+//-----------------------------------------------------------------------------
+void Catalyst2AnalysisAdaptor::AddPythonScriptPipeline(const std::string& fileName)
+{
+  conduit_cpp::Node node;
+  node["catalyst/scripts/script"].set_string(fileName);
+  auto error_code = catalyst_initialize(conduit_cpp::c_node(&node));
+  if (error_code != catalyst_status_ok)
+  {
+    // you are in trouble young man
+    std::cerr << "catalyst initialize failed with code: " << error_code << std::endl;
+  }
+}
+
+//----------------------------------------------------------------------------
+bool Catalyst2AnalysisAdaptor::Execute(DataAdaptor* dataAdaptor, DataAdaptor**)
+{
+  TimeEvent<128> mark("Catalyst2AnalysisAdaptor::Execute");
+
+  // Get a description of the simulation metadata
+  unsigned int nMeshes = 0;
+  if (dataAdaptor->GetNumberOfMeshes(nMeshes))
+  {
+    SENSEI_ERROR("Failed to get the number of meshes")
+      return false;
+  }
+
+  std::vector<MeshMetadataPtr> metadata(nMeshes);
+  for (unsigned int i = 0; i < nMeshes; ++i)
+  {
+    MeshMetadataPtr mmd = MeshMetadata::New();
+
+    if (dataAdaptor->GetMeshMetadata(i, mmd))
+    {
+      SENSEI_ERROR("Failed to get metadata for mesh " << i << " of " << nMeshes)
+        return false;
+    }
+    metadata[i] = mmd;
+  }
+
+  // Conduit node to fill
+  conduit_cpp::Node exec_params;
+
+  for (auto meta : metadata)
+  {
+    const char* meshName = meta->MeshName.c_str();
+    svtkDataObject* dobj = nullptr;
+    if (dataAdaptor->GetMesh(meshName, false, dobj))
+    {
+      SENSEI_ERROR("Failed to get mesh \"" << meshName << "\"");
+        return -1;
+    }
+
+    for (int j = 0; j < meta->NumArrays; ++j)
+    {
+      int assoc = meta->ArrayCentering[j];
+      const char *arrayName = meta->ArrayName[j].c_str();
+
+      if (dataAdaptor->AddArray(dobj, meshName, assoc, arrayName))
+      {
+        SENSEI_ERROR("Failed to add "
+                     << SVTKUtils::GetAttributesName(assoc)
+                     << " data array \"" << arrayName << "\" to mesh \""
+                     << meshName << "\"");
+          return -1;
+      }
+    }
+
+    if (dobj && std::string(meshName) != "ucdmesh")
+    {
+      if (auto* mbds = svtkMultiBlockDataSet::SafeDownCast(dobj))
```

Comment:
this will silently do nothing if the simulation does not pass a multiblock. for historical reasons simulation may pass leaf dataset types. Instead you can use `sensei::SVTKUtils::AsCompositeData` to ensure that you have a composite data set that you can iterate over. If the simulation provides a composite dataset then it is a noop, otherwise the data object is wrapped as a block of a multiblock and a smart pointer is used to clean up the multiblock.

Reply:
Great. Thanks for the tip.